metabase/metabase
# Add custom warning message when leaving editing an existing SQL question

Closes #33792

### Description

- I recommend hiding whitespace changes for the review.
- This PR only handles editing existing native questions.
- Clicking breadcrumb in query editor ([video](https://github.com/metabase/metabase/assets/6830683/74705f20-f001-43d0-b6d3-9ee54f8dbb65)) is not handled (it's edge case 5 in #34114).
- I introduced `useCallbackEffect` and refactored code from my [previous PR](https://github.com/metabase/metabase/pull/34192) to use it.

### How to verify

**Custom warning modal is shown when leaving with unsaved changes**

1. Create a native question and save it
2. Open it
3. Make some changes to the query
4. Navigate away from the page (e.g. via browser history, or links in sidebar, etc.)

Expected result: custom warning modal **is shown**.

**Custom warning modal is not shown when leaving unmodified question**

1. Create a native question and save it
2. Open it
3. Navigate away from the page (e.g. via browser history, or links in sidebar, etc.)

Expected result: custom warning modal **is not shown**.

**Custom warning modal is not shown when running modified question**

1. Create a native question and save it
2. Open it
3. Make some changes to the query
4. Run the modified question (this will cause URL to change)

Expected result: custom warning modal **is not shown**.

**Custom warning modal is not shown when saving a modified question**

1. Create a native question and save it
2. Open it
3. Make some changes to the query
4. Save the question (either replace existing question or create a new question)

Expected result: custom warning modal **is not shown**.

### Demo

https://github.com/metabase/metabase/assets/6830683/ea9e6318-23a3-4687-a3b2-4215ce08e93e
